java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -337,19 +394,56 @@ public static SlidingWindowState create(
                     listState,
                     outFieldNameToLeftTimestampIdx,
                     maxRegisteredTimer,
+                    maxTriggeredTimer,
                     lastOutputRow,
                     accumulatorStates);
         }
 
         /**
-         * Add the row to the state.
+         * Add the row to the state and adjust left idx accordingly.
          *
          * @param timestamp The row time of the row.
-         * @param row The row to be added
+         * @param row The row to be added.
+         * @param affectedAggFieldIndices The indices of the agg fields that would add or have added
+         *     this row to their accumulator. If this value is null, it means that all agg fields
+         *     add this row.
          */
-        public void addRow(long timestamp, Row row) throws Exception {
-            timestampToRow.put(timestamp, row);
-            timestampList.add(timestamp);
+        public void addRow(long timestamp, Row row, Set<Integer> affectedAggFieldIndices)
+                throws Exception {
+            if (timestampToRow.contains(timestamp)) {
+                List<Row> rows = timestampToRow.get(timestamp);
+                rows.add(row);
+                timestampToRow.put(timestamp, rows);
+            } else {
+                timestampToRow.put(timestamp, new ArrayList<>(Collections.singletonList(row)));
+
+                List<Long> timestamps = getTimestampList();
+
+                if (timestamps.isEmpty() || timestamps.get(timestamps.size() - 1) < timestamp) {
+                    timestampList.add(timestamp);
+                } else {
+                    for (int i = 0; i < timestamps.size(); i++) {
+                        if (timestamps.get(i) < timestamp) {
+                            continue;
+                        }
+                        timestamps.add(i, timestamp);
+                        break;
+                    }
+                    timestampList.update(timestamps);
+                }
+
+                List<Integer> leftIdxList = getLeftTimestampIdx();
+                for (AggregationFieldsDescriptor.AggregationFieldDescriptor aggFieldDescriptor :
+                        aggregationFieldsDescriptor.getAggFieldDescriptors()) {
+                    int aggFieldIdx =
+                            aggregationFieldsDescriptor.getAggFieldIdx(aggFieldDescriptor);
+                    if (affectedAggFieldIndices != null
```

Comment:
It appears that the affectedAggFieldIndices variable is used to determine whether the left index of an aggregator should be incremented by 1 or not. Instead of using this variable, we can check if the current left index of the aggregator is greater than or equal to the index of the added row. This way, we may be able to eliminate the need for the affectedAggFieldIndices variable.

Reply:
If the left index of an aggregation field is equal to the added row's index, we are unsure whether the left idx should increment by 1. `affectedAggFieldIndices` works in this situation.